### ISSUE TITLE
Unable to join servers to Active Directory

### ISSUE THREAD

**Issue by vinyar** [chef/knife-azure#254]
Bug: below command is not able to join newly created Azure server to Domain.

Command:
`
knife azure server create --azure-dns-name myserver --azure-vm-size Medium -I a699494373c04fc0bc8f2bb1389d6106__Windows-Server-2012-R2-20150916-en.us-127GB.vhd --azure-service-location 'Central US' --winrm-user myuser --winrm-password 'cleartextpassword' --bootstrap-protocol cloud-api --azure-network-name 'vnet_express_route' --azure-subnet-name 'Tenant Subnet' --no-node-verify-api-cert --node-ssl-verify-mode none --azure-vm-startup-timeout 15 --azure-domain-name 'ExpressRoute.domain.example.com' --azure-domain-passwd 'AnotherClearTextPassword' --azure-domain-user 'ExpressRoute\domainaccount' --azure-domain-ou-dn 'OU=SubFolder,OU=Folder,DC=ExpressRoute,DC=domain,DC=example,DC=com'
`

Notes:
If the command is executed with out the `--azure-domain-ou-dn ` flag, the error is the same.

Error:
`
..........FATAL: Unable to create role:BadRequest : Invalid domain join information specified. Ensure that the JoinDomain setting is specified, and that either the Credentials or Provisioning setting is specified.  See the Windows Automated Installation Kit documentation for more information.
ERROR: Failed to create the server -- exception being rescued: exception class/object expected
WARNING: Cleaning up resources...
..WARNING: Deleted created DNS: myserver
`
There is no stacktrace

**Comment by vinyar** [comment]
same run in debug mode (pem/id/etc.. scrubbed)
Please let me know if you need a specific ID inside XML


https://gist.github.com/vinyar/a58fe57c70880b2c82ee

**Comment by siddheshwar-more** [comment]
Hello, @vinyar 

We are able to reproduce this issue-

In above command you should specify `--azure-domain-user 'domainaccount' ` i.e without <domain_name>  and try again. 
Only use `domain_name>\username` format while login using RDP client. 

OR

Don't specify  `--azure-domain-name 'ExpressRoute.domain.example.com' ` option, as we are internally set domain name if specified `--azure-domain-user 'ExpressRoute\domainaccount' ` i.e in `domain_name>\username` format. 

Let us know your results with this solution. 

Thanks!

**Comment by adamedx** [comment]
@siddheshwar-more, thank you! Is there something we should do to make this scenario less easy to hit?

**Comment by vinyar** [comment]
Ok, we ran a few experiments with suggestions above. Outcome and feedback below:

__AD Bugs - Does join the domain with the following bugs:__
  • No support for Forest. Domain join should not attempt to infer domain based off the user name or vise versa. It could be the default behavior if the domain is not specified, however, if the domain is specified, it should be used. There can be situations where a user from a child domain has the ability to join a parent, in which case the domain is for example: Redmond.microsoft.com, but the user is Afraica\alex
  • Chef-client (with empty runlist) runs after `knife azure server create` releases hold on console. We've observed it take over 2 minutes after `server create` command finished, before the node starts showing up in `knife node list`. This operation is instant when a newly created server is not joining a domain.
__Note:__ I believe Non-AD joining `server create` keeps a lock on the console until the entire process is finished. If that's the case, lock should not be released until chef-client has finished running on the node. 

__Knife azure bug (not ad join specific):__
  • With azure server create the trusted certs are not pushed to the new box. This results in chef-client not being able to run "out of the box", or `:ssl_verify_mode verify_none` must be specified. (worked on here:   https://github.com/chef/knife-azure/issues/250)

@adamedx

**Comment by adamedx** [comment]
@vinyar, I think the new changes I specified above around naming will help the cross-forest case -- we're going to allow specification of upn's.

Regarding seeing the node appearing in Chef, that's currently by design. This is because knife bases being "finished" on the Azure "enable" status for the extension being reached. "Enable" means essentially that we started the chef service. We could have tried to trigger then "enabled" state when the node was registered, but Azure will time out if registration takes a long time. We could add a feature to poll the Chef Server for the node, though that's still a little tricky since we may not know the node name.

Regarding the last issue with trusted certs, I believe @NimishaS and team are working on that fix already.
